vpr/src/place/move_utils.cpp

```diff
@@ -589,99 +583,50 @@ const std::vector<ClusterBlockId>& movable_blocks_per_type(const t_logical_block
 
-    //check if any critical block is available
-    if (place_move_ctx.highly_crit_pins.empty()) {
+    if (movable_blocks.empty()) {
         return ClusterBlockId::INVALID();
     }
 
-    //pick a random highly critical pin and find the nets driver block
-    std::pair<ClusterNetId, int> crit_pin = place_move_ctx.highly_crit_pins[rng.irand(place_move_ctx.highly_crit_pins.size() - 1)];
-    ClusterBlockId b_from = cluster_ctx.clb_nlist.net_driver_block(crit_pin.first);
+    ClusterBlockId b_from = movable_blocks[rng.irand((int)movable_blocks.size() - 1)];
 
-    if (block_locs[b_from].is_fixed) {
-        return ClusterBlockId::INVALID(); //Block is fixed, cannot move
-    }
-
-    net_from = crit_pin.first;
-    pin_from = crit_pin.second;
     return b_from;
-
-    //Unreachable statement
-    return ClusterBlockId::INVALID();
 }
 
-//Pick a random highly critical block with a specified block type to be swapped with another random block.
-//If none is found return ClusterBlockId::INVALID()
 ClusterBlockId pick_from_highly_critical_block(ClusterNetId& net_from,
                                                int& pin_from,
                                                const int logical_blk_type_index,
                                                const PlacerState& placer_state,
+                                               const PlacerCriticalities& placer_criticalities,
                                                vtr::RngContainer& rng) {
-    auto& cluster_ctx = g_vpr_ctx.clustering();
-    auto& place_move_ctx = placer_state.move();
-    auto& block_locs = placer_state.block_locs();
+    const auto& cluster_ctx = g_vpr_ctx.clustering();
+    const auto& block_locs = placer_state.block_locs();
 
     //Initialize critical net and pin to be invalid
     net_from = ClusterNetId::INVALID();
     pin_from = -1;
 
+    const auto& highly_crit_pins = placer_criticalities.get_highly_critical_pins();
+
     //check if any critical block is available
-    if (place_move_ctx.highly_crit_pins.empty()) {
+    if (highly_crit_pins.empty()) {
         return ClusterBlockId::INVALID();
     }
 
     //pick a random highly critical pin and find the nets driver block
-    std::pair<ClusterNetId, int> crit_pin = place_move_ctx.highly_crit_pins[rng.irand(place_move_ctx.highly_crit_pins.size() - 1)];
+    std::pair<ClusterNetId, int> crit_pin = highly_crit_pins[rng.irand(highly_crit_pins.size() - 1)];
     ClusterBlockId b_from = cluster_ctx.clb_nlist.net_driver_block(crit_pin.first);
 
     //Check if picked block type matches with the blk_type specified, and it is not fixed
     //blk_type from propose move doesn't account for the EMPTY type
     auto b_from_type = cluster_ctx.clb_nlist.block_type(b_from);
-    if (b_from_type->index == logical_blk_type_index) {
+    if (b_from_type->index == logical_blk_type_index || logical_blk_type_index < 0) {
```

Comment:
I suggest adding a comment that explains what this if is doing and what logical_blk_type_index < 0 means.